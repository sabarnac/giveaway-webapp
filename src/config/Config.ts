import Participant, { Avatar } from "../store/Participant";
import ConfigJson from "./config.json";
import Random, { Engine } from "random-js";

/**
 * Interface for JSON object representing avatar details.
 */
interface AvatarJson {
  url: string;
  altText: string;
}

/**
 * Interface for JSON object representing participant details.
 */
interface ParticipantJson {
  name: string;
  avatar?: AvatarJson;
}

/**
 * Enumeration of possible animation speeds.
 */
export const enum AnimationSpeed {
  HALF = 0.5,
  ONE = 1,
  ONE_POINT_FIVE = 1.5,
  TWO = 2,
  FIVE = 5,
  TEN = 10,
  TWENTY = 20,
}

/**
 * Class representing the basic configuration of the application.
 */
class Config {
  private _randomGenerator: Random;
  private _messages: string[];
  private _allParticipants: Participant[];
  private _speed: AnimationSpeed;
  private _participantsPerMatch: AnimationSpeed;

  private getRandomGenerator = (): Random => {
    const randomEngine: Engine = Random.engines.mt19937();
    return new Random(randomEngine);
  };

  private getMessage = (): string[] => ConfigJson.messages;

  private createParticipant = (user: string | ParticipantJson): Participant =>
    typeof user === "string"
      ? new Participant(user)
      : new Participant(
          (<ParticipantJson>user).name,
          this.createAvatar((<ParticipantJson>user).avatar),
        );

  private createAvatar = (avatar?: AvatarJson): Avatar | undefined =>
    avatar ? new Avatar(avatar.url, avatar.altText) : undefined;

  private getParticipants = (): Participant[] =>
    ConfigJson.users.map(this.createParticipant);

  private getSpeed = (): AnimationSpeed => AnimationSpeed.ONE;

  private getParticipantsPerMatch = (): number =>
    ConfigJson.participantsPerMatch;

  constructor() {
    this._randomGenerator = this.getRandomGenerator();
    this._messages = this.getMessage();
    this._allParticipants = this.getParticipants();
    this._speed = this.getSpeed();
    this._participantsPerMatch = this.getParticipantsPerMatch();
  }

  /**
   * Get the random generator.
   * @return {Random}
   */
  public get randomGenerator(): Random {
    return this._randomGenerator;
  }

  /**
   * Get the list of winner/loser messages.
   * @return {string[]}
   */
  public get messages(): string[] {
    return this._messages;
  }

  /**
   * Get the list of all participants in the giveaway.
   * @return {Participant[]}
   */
  public get allParticipants(): Participant[] {
    return this._allParticipants;
  }

  /**
   * Get the speed multiplier for animations.
   * @return {AnimationSpeed}
   */
  public get speed(): AnimationSpeed {
    return this._speed;
  }

  /**
   * Set the speed multiplier for animations.
   * @param {AnimationSpeed} value The new speed multiplier value.
   */
  public set speed(value: AnimationSpeed) {
    this._speed = value;
  }

  /**
   * Get the number of participants per match.
   * @return {AnimationSpeed}
   */
  public get participantsPerMatch(): AnimationSpeed {
    return this._participantsPerMatch;
  }
}

export default new Config();
