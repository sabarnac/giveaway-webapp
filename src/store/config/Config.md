Assume the configuration in `src/config/config.json` is set as:

```json
{
  "name": "Foobar Giveaway",
  "seed": 1234567890,
  "participantsPerMatch": 2,
  "messages": [
    "#winner #loser 1",
    "#winner #loser 2",
    "#winner #loser 3",
    "#winner #loser 4"
  ],
  "users": ["FOO", "BAR", "FAZ"]
}
```

Usage:

```tsx
import AnimationSpeed from "./store/config/AnimationSpeed";
import Participant from "./store/round/match/participant/Participant";

const name: string = Config.name; // "Foobar Giveaway
const messages: string[] = Config.messages; // ["#winner #loser 1", "#winner #loser 2", "#winner #loser 3", "#winner #loser 4"]
const message: string = Config.getRandomMessage("foobar", ["barfoo"]); // "foobar barfoo 3"
const allParticipants: Participant[] = Config.allParticipants; // [Participant("FOO"), Participant("BAR"), Participant("FAZ")]
const participantsPerMatch: number = Config.participantsPerMatch; // 2
const speed: number = Config.speed; // 1
Config.setSpeed(AnimationSpeed.TWO);
```
