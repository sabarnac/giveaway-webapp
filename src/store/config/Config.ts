import Participant from "../round/match/participant/Participant";
import ConfigJson from "./config.json";
import { observable, computed, action } from "mobx";
import RandomGenerator from "./RandomGenerator";
import AnimationSpeed from "./AnimationSpeed";
import Avatar from "../round/match/participant/Avatar";

/**
 * Interface for JSON object representing avatar details.
 */
interface AvatarJson {
  /** The URL of the avatar. */
  url: string;
  /** The alternate text of the avatar image. */
  altText: string;
}

/**
 * Interface for JSON object representing participant details.
 */
interface ParticipantJson {
  /** The name of the participant. */
  name: string;
  /** The avatar of the participant (if present). */
  avatar?: AvatarJson;
}

/**
 * Class representing the basic configuration of the application.
 */
export default class Config {
  /** The name of the tournament. */
  @observable private _name: string;
  /** A list of all messages. */
  @observable private _messages: string[];
  /** The list of participants in the tournament. */
  @observable private _allParticipants: Participant[];
  /** The number of participants per match. */
  @observable private _participantsPerMatch: number;
  /** The animation speed multiplier. */
  @observable private _speed: number;
  /** The list of unused messages. */
  private _unusedMessages: string[];
  /** The list of used messages. */
  private _usedMessages: string[];

  /** The singleton instance of the class, or null if not yet created. */
  private static _instance: Config | null = null;

  private _getName = (): string => ConfigJson.name;

  private _getMessages = (): string[] => ConfigJson.messages;

  private _createParticipant = (user: string | ParticipantJson): Participant =>
    typeof user === "string"
      ? new Participant(user)
      : new Participant(
          (<ParticipantJson>user).name,
          this._createAvatar((<ParticipantJson>user).avatar)
        );

  private _createAvatar = (avatar?: AvatarJson): Avatar | undefined =>
    avatar ? new Avatar(avatar.url, avatar.altText) : undefined;

  private _getParticipants = (): Participant[] =>
    ConfigJson.users.map(this._createParticipant);

  private _getParticipantsPerMatch = (): number =>
    ConfigJson.participantsPerMatch;

  private _getSpeed = (): number => AnimationSpeed.get("ONE") as number;

  private _shouldAppend = (index: number, participantSize: number): boolean =>
    index === participantSize - 1 && participantSize > 1;

  private _appendToLastParticipant = (
    participantSize: number,
    stringToAppend: string
  ): ((name: string, index: number) => string) => (
    name: string,
    index: number
  ): string =>
    this._shouldAppend(index, participantSize)
      ? `${stringToAppend} ${name}`
      : name;

  private _formatLosers = (names: string[]): string =>
    names.map(this._appendToLastParticipant(names.length, "and")).join(", ");

  private constructor() {
    this._name = this._getName();
    this._messages = this._getMessages();
    this._unusedMessages = RandomGenerator.shuffle([...this._messages]);
    this._usedMessages = [];
    this._allParticipants = this._getParticipants();
    this._participantsPerMatch = this._getParticipantsPerMatch();
    this._speed = this._getSpeed();
  }

  /**
   * Get an instance of the config, which is a singleton.
   * @return {Config} The instance of a config.
   */
  public static getInstance(): Config {
    return Config._instance
      ? Config._instance
      : (Config._instance = new Config());
  }

  /**
   * Get the name of the tournament.
   * @return {string} The tournament name.
   */
  @computed public get name(): string {
    return this._name;
  }

  /**
   * Get the list of winner/loser messages.
   * @return {string[]} The list of mesasges.
   */
  @computed public get messages(): string[] {
    return [...this._messages];
  }

  /**
   * Gets a random message with the winner and loser.
   * @param {string} winnerName The name of the winner.
   * @param {string[]} loserNames The names of the losers.
   * @return {string} The formatted message.
   */
  public getRandomMessage(winnerName: string, loserNames: string[]): string {
    if (this._unusedMessages.length === 0) {
      this._unusedMessages = RandomGenerator.shuffle([...this._messages]);
      this._usedMessages = [];
    }
    const message = RandomGenerator.pick(this._unusedMessages);
    this._unusedMessages.splice(this._unusedMessages.indexOf(message));
    this._usedMessages.push(message);
    return message
      .replace("#winner", winnerName)
      .replace("#loser", this._formatLosers(loserNames));
  }

  /**
   * Get the list of all participants in the tournament.
   * @return {Participant[]} The list of all participants.
   */
  @computed public get allParticipants(): Participant[] {
    return this._allParticipants;
  }

  /**
   * Get the number of participants per match.
   * @return {number} The number of participants per match.
   */
  @computed public get participantsPerMatch(): number {
    return this._participantsPerMatch;
  }

  /**
   * Get the speed multiplier for animations.
   * @return {number} The speed multiplier value.
   */
  @computed public get speed(): number {
    return this._speed;
  }

  /**
   * Set the speed multiplier for animations.
   * @param {number} value The new speed multiplier value.
   */
  @action public setSpeed(value: number): void {
    if (AnimationSpeed.hasValue(value)) {
      this._speed = value;
    } else {
      console.error(`Illegal value '${value}' for speed`);
    }
  }
}
