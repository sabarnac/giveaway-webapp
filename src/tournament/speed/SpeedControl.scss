.speed-control {
  position: fixed;
  padding: 0.1rem 0.5rem;
  z-index: 1000000;
  top: 0;
  left: 0;
}
