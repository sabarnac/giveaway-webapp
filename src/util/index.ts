export let isDevEnvironment = (): boolean =>
  process.env.NODE_ENV === "development";

export let createAvatarImage = (identifier: string): string => {
  return `https://api.adorable.io/avatars/480/${identifier}`;
};

export let isInRange = (num: number, start: number, end: number): boolean =>
  num >= start && num <= end;
