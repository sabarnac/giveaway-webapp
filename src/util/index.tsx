import React, { ComponentType } from "react";
import { Dispatch, useState } from "react";
import Config from "../store/config/Config";
import { Redirect } from "react-router";
import { Observer } from "mobx-react";

/**
 * Returns if it is currently a dev environment.
 * @return {boolean} Whether it is currently a dev environment.
 */
export const isDevEnvironment = (): boolean =>
  process.env.NODE_ENV === "development";

/**
 * Returns a random avatar image URL.
 * @param {string} identifier The identifier to generate the avatar URL against.
 * @return {string} The avatar URL.
 */
export const createAvatarImage = (identifier: string): string => {
  return `https://api.adorable.io/avatars/480/${identifier}`;
};

/**
 * Checks if the number is within the range.
 * @param {number} num The number to check.
 * @param {number} start The start of the range.
 * @param {number} end The end of the range.
 * @return {boolean} Whether the number is within the range.
 */
export const isInRange = (num: number, start: number, end: number): boolean =>
  num >= start && num <= end;

/**
 * Type for custom animation state hook.
 */
export type AnimationStateHook = [
  number,
  () => void,
  (delay: number) => void,
  (state: number) => void
];

/**
 * Type for show overlay state hook.
 */
export type ShowOverlayHook = [boolean, (newValue: boolean) => void];

/**
 * Type for void function.
 */
export type VoidFunction = () => void;

/**
 * Runs the given action at least after the given delay.
 * @param {Function} action The action to run.
 * @param {number} delay The minimum delay to wait for.
 */
export const runOnDelay = (action: Function, delay: number): number =>
  setTimeout(action, delay);

/**
 * Create an animation state hook.
 * @return {AnimationStateHook} The current animation state, a state update method, and a delayed state update method.
 */
export const useAnimationState = (start: number = 0): AnimationStateHook => {
  const [currentState, setState]: [number, Dispatch<number>] = useState(start);

  const updateState = (): void => setState(currentState + 1);
  const updateStateDelay = (delay: number): VoidFunction => {
    const delayId: number = runOnDelay(updateState, delay);
    return () => clearTimeout(delayId);
  };

  return [currentState, updateState, updateStateDelay, setState];
};

/**
 * Normalize the given animation time by the animation speed multiplier.
 * @param {number} time The animation time to normalize.
 * @return {number} The normalized animation speed
 */
export const getNormalizedSpeed = (time: number): number =>
  time / Config.getInstance().speed;

/**
 * Runs the given action if the predicate is true.
 * @param {boolean} predicate The predicate to check.
 * @param {Function} action The action to run.
 */
export const runOnPredicate = (
  predicate: boolean,
  action: Function,
): VoidFunction | (() => VoidFunction) => (): void | VoidFunction =>
  predicate ? action() : undefined;

/**
 * Returns a redirect component for a given round and match.
 * @param {string} roundId The ID of the round to redirect to.
 * @param {string} matchId The ID of the match to redirect to.
 * @return {JSX.Element} The redirect element.
 */
export const getMatchRedirect = (roundId?: string, matchId?: string) => (
  <Redirect to={`/round/${roundId}/match/${matchId}`} />
);

/**
 * Returns a HOC with the given component wrapped by a MobX Observer component.
 * @param {ComponentType<T>} Component The component that should be under an observer.
 * @returns {(props: T) => JSX.Element} A HOC that wraps the given component under an observer.
 */
export const createObserver = <T extends {}>(
  Component: ComponentType<T>,
): ((props: T) => JSX.Element) => (props: T): JSX.Element => (
  <Observer>{() => <Component {...props} />}</Observer>
);
