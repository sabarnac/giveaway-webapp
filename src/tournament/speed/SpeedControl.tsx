import React, { Component } from "react";
import { observer, Observer, inject } from "mobx-react";
import "./SpeedControl.scss";
import classNames from "classnames";
import Config from "../../store/config/Config";
import AnimationSpeed from "../../store/config/AnimationSpeed";

/**
 * Properties of the animation speed controller view React component.
 */
interface SpeedControlProps {
  /** @ignore The application config. */
  config?: Config;
}

/**
 * React component for the animation speed controller.
 */
@inject("config")
@observer
export default class SpeedControl extends Component<SpeedControlProps> {
  /**
   * Sets the new animation speed.
   * @param {number} speed The new animation speed.
   */
  private setSpeed = (speed: number): void =>
    this.props.config!.setSpeed(speed);

  /**
   * Returns whether the given animation speed is the current animation speed.
   * @param {number} speed The animation speed to check.
   * @return {boolean} Whether the given animation speed is the current animation speed.
   */
  private isCurrentSpeed = (speed: number): boolean =>
    speed === this.props.config!.speed;

  /**
   * Returns the option view for the given animation speed.
   * @param {number} speed The animation speed.
   * @return {JSX.Element} The animation speed option.
   */
  private getOption = (speed: number): JSX.Element => (
    <Observer
      key={speed}
      render={() => (
        <button
          key={speed}
          className={classNames("speed-control__option", {
            "button-primary": this.isCurrentSpeed(speed)
          })}
          onClick={() => this.setSpeed(speed)}
        >
          {`${speed}X`}
        </button>
      )}
    />
  );

  /**
   * Returns the list of options for all animation speeds.
   * @return {JSX.Element[]} The list of animation speed option views.
   */
  private getOptions = (): JSX.Element[] =>
    AnimationSpeed.getValues().map(this.getOption);

  /**
   * Renders the component.
   * @return {JSX.Element} The rendered component.
   */
  public render = (): JSX.Element => (
    <div className={classNames("speed-control")}>{this.getOptions()}</div>
  );
}
