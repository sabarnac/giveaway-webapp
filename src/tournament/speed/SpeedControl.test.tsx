import React from "react";
import SpeedControl from "./SpeedControl";
import { create, ReactTestRenderer } from "react-test-renderer";
import { createDummyConfig } from "../../util/test";
import Config from "../../store/config/Config";

describe("SpeedControl Component.", () => {
  it("Matches snapshot.", () => {
    const config: Config = createDummyConfig();
    const component: ReactTestRenderer = create(
      <SpeedControl config={config} />
    );

    expect(component.toJSON()).toMatchSnapshot();
  });
});
