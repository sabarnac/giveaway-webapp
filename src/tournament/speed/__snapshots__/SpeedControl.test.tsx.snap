// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpeedControl Component. Matches snapshot. 1`] = `
<div
  className="speed-control"
>
  <button
    className="speed-control__option button-primary"
    onClick={[Function]}
  >
    1X
  </button>
  <button
    className="speed-control__option"
    onClick={[Function]}
  >
    2X
  </button>
  <button
    className="speed-control__option"
    onClick={[Function]}
  >
    3X
  </button>
  <button
    className="speed-control__option"
    onClick={[Function]}
  >
    4X
  </button>
  <button
    className="speed-control__option"
    onClick={[Function]}
  >
    5X
  </button>
</div>
`;
